File: problems/773_sliding_puzzle/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int slidingPuzzle(vector<vector<int>>& board) {
        string target = "123450", start;
        for (auto& row : board) for (int v : row) start += to_string(v);
        if (start == target) return 0;
        vector<vector<int>> moves = {{1,3},{0,2,4},{1,5},{0,4},{1,3,5},{2,4}};
        queue<string> q;
        unordered_set<string> visited;
        q.push(start); visited.insert(start);
        int step = 0;
        while (!q.empty()) {
            step++;
            int sz = q.size();
            for (int s = 0; s < sz; s++) {
                string state = q.front(); q.pop();
                int idx = state.find('0');
                for (int nei : moves[idx]) {
                    string next = state;
                    swap(next[idx], next[nei]);
                    if (next == target) return step;
                    if (!visited.count(next)) { visited.insert(next); q.push(next); }
                }
            }
        }
        return -1;
    }
};

File: problems/773_sliding_puzzle/solution.py
from typing import List


class Solution:
    def slidingPuzzle(self, board: List[List[int]]) -> int:
        target = '123450'
        start = ''.join(str(board[i][j]) for i in range(2) for j in range(3))
        if start == target:
            return 0
        moves = {0:[1,3], 1:[0,2,4], 2:[1,5], 3:[0,4], 4:[1,3,5], 5:[2,4]}
        queue = collections.deque([(start, 0)])
        visited = {start}
        while queue:
            state, step = queue.popleft()
            idx = state.index('0')
            for nei in moves[idx]:
                new_state = list(state)
                new_state[idx], new_state[nei] = new_state[nei], new_state[idx]
                new_str = ''.join(new_state)
                if new_str == target:
                    return step + 1
                if new_str not in visited:
                    visited.add(new_str)
                    queue.append((new_str, step + 1))
        return -1
